Error while checking build/O2/fullCI_slc9 for 7b5264a60c280c7662097a59fbb17b545582eeea at 2025-06-17 10:44:
```
## sw/BUILD/O2-latest/log
/sw/SOURCES/O2/14397-slc9_x86-64/0/GPU/GPUTracking/Global/GPUErrors.h:20:10: fatal error: 'unordered_map' file not found
/sw/SOURCES/O2/14397-slc9_x86-64/0/GPU/GPUTracking/Global/GPUErrors.h:20:10: fatal error: unordered_map: No such file or directory
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14397/7b5264a60c280c7662097a59fbb17b545582eeea/build_O2_fullCI_slc9/pretty.html).

--- a/Detectors/TPC/qc/include/TPCQC/GPUErrorQA.h
+++ b/Detectors/TPC/qc/include/TPCQC/GPUErrorQA.h
@@ -54,7 +54,7 @@ class GPUErrorQA
   void resetHistograms();
 
   /// return histograms
-  const std::unordered_map<std::string, std::unique_ptr<TH1>>& getMapHist() const { return mMapHist; };
+  const std::unordered_map<std::string, std::unique_ptr<TH1>>& getMapHist() const { return mMapHist; }
 
   /// Dump results to a file
   void dumpToFile(std::string filename);
@@ -62,7 +62,7 @@ class GPUErrorQA
  private:
   std::unordered_map<std::string, std::unique_ptr<TH1>> mMapHist;
 
-  ClassDefNV(GPUErrorQA, 1);
+  ClassDefNV(GPUErrorQA, 2);
 };
 } // namespace o2::tpc::qc
 

--- a/Detectors/TPC/qc/include/TPCQC/GPUErrorQA.h
+++ b/Detectors/TPC/qc/include/TPCQC/GPUErrorQA.h
@@ -18,22 +18,20 @@
 #define AliceO2_TPC_QC_GPUERRORQA_H
 
 #include <memory>
-#include <gsl/span>
+#include <string>
+#include <vector>
+#include <unordered_map>
 
 // root includes
-#include "TH1.h"
 
 // o2 includes
 // #include "DataFormatsTPC/Defs.h"
 
-namespace o2
-{
-namespace tpc
-{
-namespace qc
+class TH1;
+namespace o2::tpc::qc
 {
 
-/// @brief  TPC QC task for errors from GPU reconstruction
+/// @brief TPC QC task for errors from GPU reconstruction
 ///
 /// This class is used to retrieve and visualize GPU errors
 /// according to corresponding error code and location.
@@ -47,7 +45,7 @@ class GPUErrorQA
   GPUErrorQA() = default;
 
   /// process gpu error reported by the reconstruction workflow
-  void processErrors(gsl::span<const std::array<uint32_t, 4>> errors);
+  void processErrors(std::vector<std::array<uint32_t, 4>> errors);
 
   /// Initialize all histograms
   void initializeHistograms();
@@ -55,15 +53,17 @@ class GPUErrorQA
   /// Reset all histograms
   void resetHistograms();
 
+  /// return histograms
+  const std::unordered_map<std::string, std::unique_ptr<TH1>>& getMapHist() const { return mMapHist; };
+
   /// Dump results to a file
   void dumpToFile(std::string filename);
 
  private:
-  std::unique_ptr<TH1F> mHist;
-  ClassDefNV(GPUErrorQA, 1)
+  std::unordered_map<std::string, std::unique_ptr<TH1>> mMapHist;
+
+  ClassDefNV(GPUErrorQA, 1);
 };
-} // namespace qc
-} // namespace tpc
-} // namespace o2
+} // namespace o2::tpc::qc
 
 #endif // AliceO2_TPC_QC_GPUERRORQA_H

--- a/Detectors/TPC/qc/src/GPUErrorQA.cxx
+++ b/Detectors/TPC/qc/src/GPUErrorQA.cxx
@@ -17,7 +17,7 @@
 
 // o2 includes
 #include "TPCQC/GPUErrorQA.h"
-#include "GPUDefMacros.h"
+#include "GPUErrors.h"
 
 ClassImp(o2::tpc::qc::GPUErrorQA);
 
@@ -28,21 +28,13 @@ void GPUErrorQA::initializeHistograms()
 {
   TH1::AddDirectory(false);
 
-  // get gpu error names
-  // copied from GPUErrors.h
-  static std::unordered_map<uint32_t, const char*> errorNames = {
-#define GPUCA_ERROR_CODE(num, name, ...) {num, GPUCA_M_STR(name)},
-#include "GPUErrorCodes.h"
-#undef GPUCA_ERROR_CODE
-  };
-
   // 1D histogram counting all reported errors
-  mMapHist["ErrorCounter"] = std::make_unique<TH1I>("ErrorCounter", "ErrorCounter", errorNames.size(), -0.5, errorNames.size() - 0.5);
+  mMapHist["ErrorCounter"] = std::make_unique<TH1I>("ErrorCounter", "ErrorCounter", o2::gpu::errorNames.size(), -0.5, o2::gpu::errorNames.size() - 0.5);
   mMapHist["ErrorCounter"]->GetXaxis()->SetTitle("Error Codes");
   mMapHist["ErrorCounter"]->GetYaxis()->SetTitle("Entries");
   // for convienence, label each bin with the error name
   for (size_t bin = 1; bin < mMapHist["ErrorCounter"]->GetNbinsX(); bin++) {
-    auto const& it = errorNames.find(bin);
+    auto const& it = o2::gpu::errorNames.find(bin);
     mMapHist["ErrorCounter"]->GetXaxis()->SetBinLabel(bin, it->second);
   }
 }
@@ -68,7 +60,7 @@ void GPUErrorQA::dumpToFile(const std::string filename)
   auto f = std::unique_ptr<TFile>(TFile::Open(filename.data(), "recreate"));
   TObjArray arr;
   arr.SetName("GPUErrorQA_Hists");
-  for (const auto& [name, hist] : mMapHist) {
+  for ([[maybe_unused]] const auto& [name, hist] : mMapHist) {
     arr.Add(hist.get());
   }
   arr.Write(arr.GetName(), TObject::kSingleKey);

--- a/Detectors/TPC/qc/src/GPUErrorQA.cxx
+++ b/Detectors/TPC/qc/src/GPUErrorQA.cxx
@@ -11,16 +11,13 @@
 
 #define _USE_MATH_DEFINES
 
-#include <cmath>
-#include <memory>
-
 // root includes
 #include "TFile.h"
-#include <TH1.h>
+#include "TH1I.h"
 
 // o2 includes
 #include "TPCQC/GPUErrorQA.h"
-#include "GPUErrors.h"
+#include "GPUDefMacros.h"
 
 ClassImp(o2::tpc::qc::GPUErrorQA);
 
@@ -30,26 +27,49 @@ using namespace o2::tpc::qc;
 void GPUErrorQA::initializeHistograms()
 {
   TH1::AddDirectory(false);
-  mHist = std::make_unique<TH1F>("ErrorCounter", "ErrorCounter", o2::gpu::GPUErrors::getMaxErrors(), 0, o2::gpu::GPUErrors::getMaxErrors());
+
+  // get gpu error names
+  // copied from GPUErrors.h
+  static std::unordered_map<uint32_t, const char*> errorNames = {
+#define GPUCA_ERROR_CODE(num, name, ...) {num, GPUCA_M_STR(name)},
+#include "GPUErrorCodes.h"
+#undef GPUCA_ERROR_CODE
+  };
+
+  // 1D histogram counting all reported errors
+  mMapHist["ErrorCounter"] = std::make_unique<TH1I>("ErrorCounter", "ErrorCounter", errorNames.size(), -0.5, errorNames.size() - 0.5);
+  mMapHist["ErrorCounter"]->GetXaxis()->SetTitle("Error Codes");
+  mMapHist["ErrorCounter"]->GetYaxis()->SetTitle("Entries");
+  // for convienence, label each bin with the error name
+  for (size_t bin = 1; bin < mMapHist["ErrorCounter"]->GetNbinsX(); bin++) {
+    auto const& it = errorNames.find(bin);
+    mMapHist["ErrorCounter"]->GetXaxis()->SetBinLabel(bin, it->second);
+  }
 }
 //______________________________________________________________________________
 void GPUErrorQA::resetHistograms()
 {
-  mHist->Reset();
+  for (const auto& pair : mMapHist) {
+    pair.second->Reset();
+  }
 }
 //______________________________________________________________________________
-void GPUErrorQA::processErrors(gsl::span<const std::array<uint32_t, 4>> errors)
+void GPUErrorQA::processErrors(std::vector<std::array<uint32_t, 4>> errors)
 {
   for (const auto& error : errors) {
     uint32_t errorCode = error[0];
-    mHist->Fill(static_cast<float>(errorCode));
+    mMapHist["ErrorCounter"]->AddBinContent(errorCode);
   }
 }
 
 //______________________________________________________________________________
 void GPUErrorQA::dumpToFile(const std::string filename)
 {
-  auto f = std::unique_ptr<TFile>(TFile::Open(filename.c_str(), "recreate"));
-  mHist->Write();
-  f->Close();
+  auto f = std::unique_ptr<TFile>(TFile::Open(filename.data(), "recreate"));
+  TObjArray arr;
+  arr.SetName("GPUErrorQA_Hists");
+  for (const auto& [name, hist] : mMapHist) {
+    arr.Add(hist.get());
+  }
+  arr.Write(arr.GetName(), TObject::kSingleKey);
 }

--- a/GPU/GPUTracking/Global/GPUErrors.cxx
+++ b/GPU/GPUTracking/Global/GPUErrors.cxx
@@ -36,7 +36,6 @@ GPUd() void GPUErrors::raiseError(uint32_t code, uint32_t param1, uint32_t param
 #ifndef GPUCA_GPUCODE
 
 #include <cstring>
-#include <unordered_map>
 
 uint32_t GPUErrors::getMaxErrors()
 {
@@ -47,12 +46,6 @@ void GPUErrors::clear()
 {
   memset(mErrors, 0, GPUCA_MAX_ERRORS * sizeof(*mErrors));
 }
-
-static std::unordered_map<uint32_t, const char*> errorNames = {
-#define GPUCA_ERROR_CODE(num, name, ...) {num, GPUCA_M_STR(name)},
-#include "GPUErrorCodes.h"
-#undef GPUCA_ERROR_CODE
-};
 
 bool GPUErrors::printErrors(bool silent, uint64_t mask)
 {

--- a/GPU/GPUTracking/Global/GPUErrors.cxx
+++ b/GPU/GPUTracking/Global/GPUErrors.cxx
@@ -36,6 +36,7 @@ GPUd() void GPUErrors::raiseError(uint32_t code, uint32_t param1, uint32_t param
 #ifndef GPUCA_GPUCODE
 
 #include <cstring>
+#include <unordered_map>
 
 uint32_t GPUErrors::getMaxErrors()
 {
@@ -47,9 +48,20 @@ void GPUErrors::clear()
   memset(mErrors, 0, GPUCA_MAX_ERRORS * sizeof(*mErrors));
 }
 
+const std::unordered_map<uint32_t, const char*>& GPUErrors::getErrorNames()
+{
+  static std::unordered_map<uint32_t, const char*> errorNames = {
+#define GPUCA_ERROR_CODE(num, name, ...) {num, GPUCA_M_STR(name)},
+#include "GPUErrorCodes.h"
+#undef GPUCA_ERROR_CODE
+  };
+  return errorNames;
+}
+
 bool GPUErrors::printErrors(bool silent, uint64_t mask)
 {
   bool retVal = 0;
+  const auto& errorNames = getErrorNames();
   for (uint32_t i = 0; i < std::min(*mErrors, GPUCA_MAX_ERRORS); i++) {
     uint32_t errorCode = mErrors[4 * i + 1];
     const auto& it = errorNames.find(errorCode);

--- a/GPU/GPUTracking/Global/GPUErrors.h
+++ b/GPU/GPUTracking/Global/GPUErrors.h
@@ -16,9 +16,17 @@
 #define GPUERRORS_H
 
 #include "GPUCommonDef.h"
+#include "GPUDefMacros.h"
+#include <unordered_map>
 
 namespace o2::gpu
 {
+
+static std::unordered_map<uint32_t, const char*> errorNames = {
+#define GPUCA_ERROR_CODE(num, name, ...) {num, GPUCA_M_STR(name)},
+#include "GPUErrorCodes.h"
+#undef GPUCA_ERROR_CODE
+};
 
 class GPUErrors
 {

--- a/GPU/GPUTracking/Global/GPUErrors.h
+++ b/GPU/GPUTracking/Global/GPUErrors.h
@@ -16,17 +16,13 @@
 #define GPUERRORS_H
 
 #include "GPUCommonDef.h"
-#include "GPUDefMacros.h"
+
+#ifndef GPUCA_GPUCODE
 #include <unordered_map>
+#endif
 
 namespace o2::gpu
 {
-
-static std::unordered_map<uint32_t, const char*> errorNames = {
-#define GPUCA_ERROR_CODE(num, name, ...) {num, GPUCA_M_STR(name)},
-#include "GPUErrorCodes.h"
-#undef GPUCA_ERROR_CODE
-};
 
 class GPUErrors
 {
@@ -42,6 +38,7 @@ class GPUErrors
   void setMemory(GPUglobalref() uint32_t* m) { mErrors = m; }
   void clear();
   bool printErrors(bool silent = false, uint64_t mask = 0);
+  static const std::unordered_map<uint32_t, const char*>& getErrorNames();
   uint32_t getNErrors() const;
   const uint32_t* getErrorPtr() const;
   static uint32_t getMaxErrors();

